### PR TITLE
Update alias design with additional notes

### LINKED ISSUE #3827: Design alias to public key validation logic
### Problem

As part of the expansion of the expansion of the alias transaction applicability we will need to add logic to ensure an alias maps to the public key of the account during set.
No such design exists

### Solution

Update the alias transaction applicability expansion design doc with an approach to

- ensure a provided alias matches expected input i.e. `ED25519`/`ECDSA` public key or `evm_address` style
- ensue no matching alias exists
- ensure alias either equals (when public key) or can be extracted from public key provided (when `evm_address`)

### Alternatives

_No response_

### LINKED ISSUE #3828: Design public key or ethereum address alias configuration
### Problem

As part of the expansion of the expansion of the alias transaction applicability we will need to add logic to support the setting of an alias during a transaction
No such design exists

### Solution

Update the alias transaction applicability expansion design doc with an approach to

- take in the aliasBytes and confirm validity 
- support ED25519/ECDSA public key uniqueness check
- setting of alias and ledger update

### Alternatives

_No response_

### LINKED ISSUE #3829: Design lazy create on accounts when ethereum address provided
### Problem

As part of the expansion of the expansion of the alias transaction applicability we will need to add logic to support the handling of ethereum account address as a provided alias
No such design exists

### Solution

Update the alias transaction applicability expansion design doc with an approach to

- take in ethereum address i.e. confirm length and format
- set alias as this value
- updated ledger
- ensure 
- support extracting a public key from a future transaction
- check that ethereum address of public key provided matches ethereum address alias set prior
- ensure account is functional

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by MiroslavGatsanoga**
Signed-off-by: Miroslav Gatsanoga <miroslav.gatsanoga@limechain.tech>

**Description**:

**Related issue(s)**: **Related issue(s)**: https://github.com/hashgraph/hedera-services/issues/3827, https://github.com/hashgraph/hedera-services/issues/3828, https://github.com/hashgraph/hedera-services/issues/3829

Fixes #3827
Fixes #3828 
Fixes #3829 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: docs/design/alias-configuration.md [hunk 1]
```diff
@@ -27,10 +27,27 @@ This document explains how additional Crypto transaction flows may be used to cr
...
+
+- For CryptoTransfer we already support alias public key auto create
+- To support Ethereum address for CryptoTransfer we will need to: 
+  - Change the implementation in AliasResolver.resolveInternalFungible method, it already has some EVM address handling
+  - Change the AutoCreationLogic.create to expect alias that might not be key but a public address
 
 ### Crypto Create Transaction Alias Logic
 
+- Extract alias from CryptoCreateTransactionBody and validate it in validate() method, checking if the alias matches the key or can be derived from the key (for EVM address alias)
+- For the ED25519 key case when creating the account we should set its alias to the value of the key
```
Comment (Nana-EC) [inline docs/design/alias-configuration.md:47]: We don't want to default set the alias for ED25519 case

File: docs/design/alias-configuration.md [hunk 2]
```diff
@@ -27,10 +27,27 @@ This document explains how additional Crypto transaction flows may be used to cr
...
+- For the ECDSA key case when creating the account we should set its alias to the value of the key as well
```
Comment (Nana-EC) [inline docs/design/alias-configuration.md:48]: Add note for ensuring the alias -> accountId map is populated with the evmAddress for immediate use after creation

File: docs/design/alias-configuration.md (hunk 1 above)
Comment (Nana-EC) [inline docs/design/alias-configuration.md:47]: We don't want to default set the alias for ED25519 case

File: docs/design/alias-configuration.md [hunk 3]
```diff
@@ -52,4 +69,56 @@ In this flow
...
 ## Open Questions
 
-## Acceptance Tests
\ No newline at end of file
+## Acceptance Tests
+
+## Cases
+
+- Crypto Transfer Transaction to EVM address alias should create an account with its alias field set to the EVM address and its key will be empty (to be populated by a signed transaction as part of lazy account create)
+- Crypto Transfer Transaction (without auto creation): to be called with EVM address (the account with this address is already auto lazy created)
```
Comment (Nana-EC) [inline docs/design/alias-configuration.md:77]: This can be any transaction not just crypto transfer.
For instance is my account was lazily created.
If i issued a ContractCall and signed from metamask it should be able to pick up my signature, fill in the account and carry on to process the ContractCall.
This should be seamless

Comment (Nana-EC) [reply docs/design/alias-configuration.md:77]: So we should have cases for `CryptoTransfer`, `ContractCreate`, `ContractCall`, `EthereumTransaction`, `TokenAssociation` to begin with.
We can discuss others but I believe these represent the next logical steps one might take

File: docs/design/alias-configuration.md [hunk 4]
```diff
@@ -52,4 +69,56 @@ In this flow
...
+  - We need to extract the public key from signature with extractSig
+  - After that to use the extracted key for recovering the public address (recoverAddressFromPubKey)
+  - Recovered address must match with the alias (public address of the eth account)
+  - If the recovered address match with the address stored in the alias we have to update the key property of the account and store it in the ledger
+  - With the above step the process of lazy creation will be finished
+- Crypto Transfer Transaction to ECDSA key alias should create an account with its alias field set to the ECDSA key, populate the alias map with the EVM address calculated based on the ECDSA key, and the key field set with the value of the ECDSA key (currently implemented) and another Crypto Transfer Transaction to the EVM address corresponding to the ECDSA key should transfer to the account created in the previous transaction
+- Crypto Create Transaction with ED25519 key alias should create an account with its alias and key fields set to the ED25519 key. Attempting another Crypto Create Transaction with the same ED25519 key alias should fail
```
Comment (Nana-EC) [inline docs/design/alias-configuration.md:84]: Remove this

File: docs/design/alias-configuration.md [hunk 5]
```diff
@@ -52,4 +69,56 @@ In this flow
...
+    2. a) When a signed transaction referencing the account by the EVM_address alias is received:
+        - Account with:
+            - hedera id: 0.0.20
+            - admin key: ECDSA key
+            - alias: 0.0.ECDSA key
+        - AliasManager with:
+            - map with the EVM_address alias based on the ECDSA key → hedera id
+            - map with the 0.0.ECDSA key alias → hedera id
+    
+    2. b) Don’t overwrite the alias field, instead add logic during the map rebuilding
```
Comment (Nana-EC) [inline docs/design/alias-configuration.md:117]: Can you clarify why we wouldn't override the alias field stored tot eh public key once we get it?
Since we're trying to map into existing flow we should ensure that the only case where an alias on account (not the map) is an evmAddress is when it's a Hollow account i.e. in between state and waiting for a subsequent transaction to complete the onboarding process

File: docs/design/alias-configuration.md [hunk 6]
```diff
@@ -52,4 +69,56 @@ In this flow
...
+    - if the value of the alias field of the account has the evm_address format, add an entry to the aliases map with the value of 0.0.ECDSA key which is set on the account
+
+### Notes
+
+- For the EVM address cases we should store the alias value in the state not only in the memory structure
```
Comment (Nana-EC) [inline docs/design/alias-configuration.md:122]: See previous note. To simplify things let's not yet do this. State should only have evmAddress as an interrim approach whiles the account is not complete

File: docs/design/alias-configuration.md [hunk 7]
```diff
@@ -52,4 +69,56 @@ In this flow
...
+- It is ok for different accounts to have the same key
```
Comment (Nana-EC) [inline docs/design/alias-configuration.md:123]: This is the case today. The AccountId is the real differentiator. Even if you have multiple accounts with the same public key only one can have it as the alias

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Some thoughts

File: docs/design/alias-configuration.md (hunk 2 above)
Comment (MiroslavGatsanoga) [reply docs/design/alias-configuration.md]: Added.

File: docs/design/alias-configuration.md (hunk 3 above)
Comment (MiroslavGatsanoga) [reply docs/design/alias-configuration.md]: Updated the description of this case to be for any type of signed transaction and listed `CryptoTransfer`, `ContractCreate`, `ContractCall`, `EthereumTransaction`, `TokenAssociation` as examples.

File: docs/design/alias-configuration.md (hunk 1 above)
Comment (MiroslavGatsanoga) [reply docs/design/alias-configuration.md:47]: I am fine with removing the Crypto Create case for ED25519 key alias, but I think currently the auto create with Crypto Transfer is working with ED25519 keys. That's why I assumed that in order to be consistent we will want to have it for Crypto Create as well. What is the reason we don't want to support this?

File: docs/design/alias-configuration.md (hunk 4 above)
Comment (MiroslavGatsanoga) [reply docs/design/alias-configuration.md:84]: See https://github.com/hashgraph/hedera-services/pull/3943#discussion_r972928819.

File: docs/design/alias-configuration.md (hunk 5 above)
Comment (MiroslavGatsanoga) [reply docs/design/alias-configuration.md:117]: The 2.b) scenario was written down as an alternative to 2.a) i.e. if we are not fine with an approach in which we update the account alias. In 1. we create an account with alias set to the EVM address and empty key and then in 2.a) we set the key and update the alias to have the ECDSA key value.

We were not sure if updating the account alias is ok as an approach but if that is fine we will prefer to go with the 2.a) option.

File: docs/design/alias-configuration.md (hunk 6 above)
Comment (MiroslavGatsanoga) [reply docs/design/alias-configuration.md:122]: The idea of the note is that we are going to need to store the EVM address alias in state (as the alias of the account) and not only in the in-memory map that is rebuild from state, because if we have it only in the in-memory map and we have a restart then we will lose the entry linking the EVM to an accountId.

File: docs/design/alias-configuration.md (hunk 1 above)
Comment (Nana-EC) [reply docs/design/alias-configuration.md:47]: Yes auto-create with cryptoTransfer is the only way today.
Alias are immutable and ED account with alias don't enable any functionality whereas EC keys with alias open up smart contract compatibility. 
There's a migration path for EC but none for ED

File: docs/design/alias-configuration.md (hunk 5 above)
Comment (Nana-EC) [reply docs/design/alias-configuration.md:117]: We should be updating the account alias if it is the evm address and we get a publickey in the signature map that maps to the evm address

File: docs/design/alias-configuration.md (hunk 1 above)
Comment (MiroslavGatsanoga) [reply docs/design/alias-configuration.md]: Removed the ED25519 case.

File: docs/design/alias-configuration.md (hunk 4 above)
Comment (MiroslavGatsanoga) [reply docs/design/alias-configuration.md]: Removed.

File: docs/design/alias-configuration.md [hunk 8]
```diff
@@ -27,10 +27,26 @@ This document explains how additional Crypto transaction flows may be used to cr
 
 ### Alias to Public Key Validation
 
-### Support Public Key or Ethereum Address Alas
+- In case of ED25519 the provided alias must match the public key of the account that we want to create
+- In case of the ECDSA the alias must match the public key or be the public address of the ECDSA key provided/set on the account (when the alias is the public address of the ECDSA key and the key is not set on the account we have to create the account with empty public key field and to store the public address as alias)
+- We can use MiscUtils.isSerializedProtoKey logic to validate keys
+- Additionally we can validate with regex for evm_address (if hex encoded string), otherwise if bytes array check its length
+- We can check existing aliases using AliasManager.lookupIdBy method
```
Comment (Nana-EC) [inline docs/design/alias-configuration.md:34]: Note this may not extend long term.
Currently the Alias map includes public address but we should account for scenarios where it only contains public keys.

File: docs/design/alias-configuration.md [hunk 9]
```diff
@@ -27,10 +27,26 @@ This document explains how additional Crypto transaction flows may be used to cr
...
+- For CryptoTransfer we already support alias public key auto create
+- To support Ethereum address for CryptoTransfer we will need to: 
+  - Change the implementation in AliasResolver.resolveInternalFungible method, it already has some EVM address handling
+  - Change the AutoCreationLogic.create to expect alias that might not be key but a public address
 
 ### Crypto Create Transaction Alias Logic
 
+- Extract alias from CryptoCreateTransactionBody and validate it in validate() method, checking if the alias matches the key or can be derived from the key (for EVM address alias)
+- For the ECDSA key case when creating the account we should set its alias to the value of the key as well. We should also add entries to the alias->accountId map for the ECDSA key and the EVM address derived from it
+- For the ETH address case we can't set the key for the account (as it will be missing), that would be done as the final part of the lazy account create
```
Comment (Nana-EC) [inline docs/design/alias-configuration.md:48]: alias should be used to store ethereum Address in this case

File: docs/design/alias-configuration.md [hunk 10]
```diff
@@ -52,4 +68,55 @@ In this flow
...
+            - map with the 0.0.ECDSA key alias → hedera id
+    
+    2. b) Don’t overwrite the alias field, instead add logic during the map rebuilding
+    - if the value of the alias field of the account has the evm_address format, add an entry to the aliases map with the value of 0.0.ECDSA key which is set on the account
+
+### Notes
+
+- For the EVM address cases we should store the alias value in the state not only in the memory structure
+- It is ok for different accounts to have the same key
+- For accounts with non-empty aliases if we have a key that is set, it should match the alias (for key aliases the values should be equal and for EVM aliases the key should be the ECDSA key from which the EVM address is derived)
```
Comment (Nana-EC) [inline docs/design/alias-configuration.md:122]: Can you add a test case and design logic for 
`EthereumTransaction` where the t`x.value` is set (and therefore it's a cryptotransfer) but the ethereum adddress noted in `tx.to` is not. 
We should do a lazy create in this case and create a hollow account

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Looking good, one final suggestion.

**Comment by Nana-EC** [comment]
Also if not longer needed you can remove from draft

**Comment by Nana-EC** [comment]
Can see [HIP](https://github.com/hashgraph/hedera-improvement-proposal/pull/583) for more details now

File: docs/design/alias-configuration.md (hunk 9 above)
Comment (MiroslavGatsanoga) [reply docs/design/alias-configuration.md]: Added additional description.

File: docs/design/alias-configuration.md (hunk 8 above)
Comment (MiroslavGatsanoga) [reply docs/design/alias-configuration.md:34]: Currently the aliases map contains entries for public keys and public addresses that are derived from public keys.

For the case of the creation of hollow account with public address (to be implemented) I can see that the aliases map will contain entry only for public address and not for the corresponding public key. 
Adding the public key entry corresponding to the public address will be done as finalizing the lazy account creation, which will leave the aliases map in the same state we currently have i.e. entries with public keys and public addresses corresponding to those public keys.

What is the case we may have in the future in which we can have only public key entries?

File: docs/design/alias-configuration.md (hunk 5 above)
Comment (stoqnkpL) [reply docs/design/alias-configuration.md:115]: `2. a)` and `2. b)` were two options that we needed to discuss and choose from. 
Now that we have chosen option `2. a)` I think we should delete `2. b)` from this document for clarity and leave only `2.` with added description for the update of the account alias.

File: docs/design/alias-configuration.md (hunk 6 above)
Comment (stoqnkpL) [reply docs/design/alias-configuration.md:120]: > The idea of the note is that we are going to need to store the EVM address alias in state (as the alias of the account) and not only in the in-memory map that is rebuild from state, because if we have it only in the in-memory map and we have a restart then we will lose the entry linking the EVM to an accountId.

I think this is correct but the way it's written above:
> For the EVM address cases we should store the alias value in the state not only in the memory structure

it looks like we'll **always** store the evm_address alias, which is incorrect. I would rephrase this to something like:
When lazy-creating an account we'll temporarily store the evm_address alias in the alias field of the account. It will be replaced by the public_key alias when the account creation is complete.

File: docs/design/alias-configuration.md (hunk 10 above)
Comment (stoqnkpL) [reply docs/design/alias-configuration.md:122]: @Nana-EC I'm not sure I understand how this will work. We need at least an `evm_address` alias for the lazy created account. What would the alias be in this case, if `to` is not set in the `EthereumTransaction`?

Comment (Nana-EC) [reply docs/design/alias-configuration.md:122]: This would only refer to the case were `to` is set and `value` is non zero.
It's basically to expose account creation that we do via HAPI when sending HBAR via the EthereumTransaction to support Ethereum RPC native wallets and tools

Comment (MiroslavGatsanoga) [reply docs/design/alias-configuration.md:126]: Added additional case.

File: docs/design/alias-configuration.md (hunk 5 above)
Comment (Nana-EC) [reply docs/design/alias-configuration.md]: @MiroslavGatsanoga  and @stoqnkpL Final note on this.
Reverting my above note.
For lazy create we should not update the alias from evmAddress to publicKey.
We should only set the alias as the publicKey when a user does auto-create and submits the public key.
I'll also update the HIP with this

**Review by shemnon — APPROVED** [review]

**Review by lukelee-sl — APPROVED** [review]

**Review by Nana-EC — APPROVED** [review]
LG
You can check this in.
In follow up could you add a section for `EthereumTransacion` case that represent an HBAR transfer i.e. where `tx.to` and `tx.value` are set that results in 
- a lazy create 
- hollow account -> complete account extracting public key from inner transaction
